### PR TITLE
Added Settings Page with DefaultSyncDays settings

### PR DESCRIPTION
**Pull request by Kamlesh72**
## Linked Issue(s) 

#500 

## Acceptance Criteria fulfillment

- [x] Reducers for Getting and Updating DefaultSyncDaysSetting
- [x] NextJs common setting API for Getting and Updating Org settings.
- [x] NextJs API for Resetting Bookmark
- [x] Added Required types
- [x] Settings Page and SyncDaysSetting component

### CODE REVIEW

**Comment by Kamlesh72** [comment]
@samad-yar-khan @jayantbh Is this approach ok? The settings page still need changes to easily add further settings. We can discuss about it.

**Comment by adnanhashmi09** [comment]
@Kamlesh72 Can you also add a screenshot for the UI changes that goes with this PR. Thanks :)

**Comment by Kamlesh72** [comment]
https://github.com/user-attachments/assets/05a15de4-cf9f-4ccd-bc36-714e8e851f07

Screenshot 2024-08-12 at 3 21 41 PM

File: web-server/pages/api/internal/[org_id]/bookmarks.ts
Comment (jayantbh) [inline web-server/pages/api/internal/[org_id]/bookmarks.ts:26]: Remove any commented out code.

File: web-server/pages/api/internal/[org_id]/settings.ts
Comment (jayantbh) [inline web-server/pages/api/internal/[org_id]/settings.ts:24]: `user_id` isn't needed here, or in the schema since we don't have the concept of "Users" in this app yet.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (jayantbh) [inline web-server/src/components/Settings/SyncDaysSetting.tsx:22]: [not a required change]

```suggestion
  const syncDays = useEasyState(31);
```

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (jayantbh) [inline web-server/src/components/Settings/SyncDaysSetting.tsx:131]: This is what causes the layout to shift when the input field is in focus. I think that looks visually displeasing.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (jayantbh) [inline web-server/src/components/Settings/SyncDaysSetting.tsx:56]: I think all this should be part of the same thunk that updates sync days.
Since both of those actions needs to succeed collectively.

In any case, @samad-yar-khan believes it's based to not make the UI aware of this technical detail, and the backend should internally handle it. So this would change as soon as he's done with his change.

We'll have to wait till then.

**Review by jayantbh — COMMENTED** [review]
A lot of sync logic is component specific right now. And while right now it's fine, I think we should drive as much of the settings storage and updation from redux and thunks as possible.

Because when there's multiple settings, you don't want to maintain a variable for state and request state for each of those.

You want to be able to update/handle settings changes collectively.

Of course, you don't need to overengineer that right now, but you can set up the foundational code for making it possible.

Let me know if it feels too complex, and you can take up that effort in a subsequent PR.

**Review by jayantbh — CHANGES_REQUESTED** [review]
My last one was supposed to be a request for changes. 🤷🏽

File: web-server/pages/api/internal/[org_id]/bookmarks.ts
Comment (samad-yar-khan) [inline web-server/pages/api/internal/[org_id]/bookmarks.ts:32]: @Kamlesh72  this can be removed for now. I have incorporate this side effect in the backend for now: https://github.com/middlewarehq/middleware/pull/518. Makes things simpler. Just update the setting and handle for setting updation failure.

Comment (Kamlesh72) [reply web-server/pages/api/internal/[org_id]/bookmarks.ts:32]: Ok sure.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:131]: Yes if we make single SAVE/DISCARD button to handle all settings, this can be solved.

**Comment by Kamlesh72** [comment]
@jayantbh I will add code to handle the less than currentSyncDays input case. What should be maximum sync days? 365,366? 

I am thinking of some way to handle all settings collectively. Will share the approach soon.

**Comment by jayantbh** [comment]
I think let's not set a maximum... but, let's start by setting it to 366 days. @Kamlesh72

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (samad-yar-khan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:56]: handled in #518

File: web-server/src/types/resources.ts
Comment (samad-yar-khan) [inline web-server/src/types/resources.ts:667]: @Kamlesh72 Although not in this PR,  but would you to open a subsequent PR to  completely remove the usage of OrgAlertSettings. Its not being used in the backend anymore. Also, longer names might sound verbose but something like `OrgShouldSyncAlertsAsIncidentsSetting` would have been a better name change, makes it more obvious.

Comment (Kamlesh72) [reply web-server/src/types/resources.ts:667]: Ok sure I will make the PR. 
I will keep in mind for better naming.

**Comment by Kamlesh72** [comment]
@samad-yar-khan do I need to call syncReposForOrg() after updating Sync Days?

**Comment by samad-yar-khan** [comment]
> @samad-yar-khan do I need to call syncReposForOrg() after updating Sync Days?

@Kamlesh72  yes you should. You can also ask the customer if they want to backfill data as per current changes or let the system handle in next sync cycle.

**Comment by samad-yar-khan** [comment]
https://github.com/middlewarehq/middleware/pull/516#issuecomment-2288883466

**Comment by Kamlesh72** [comment]
@samad-yar-khan In future we might have more settings and they gets updated together. So asking customer for change might not be a good choice. I think it will be better to just call the sync in background.

**Comment by adnanhashmi09** [comment]
@Kamlesh72 I agree with @samad-yar-khan. Either the data fills itself in the next iteration, or the customer force syncs. calling it quietly in the background might not be very intuitive for the user. 
If we were so start the sync as soon as the user sets the setting, it would be better to show some message to the user that the sync has started or something along those lines. @jayantbh What are your thoughts on this?

**Comment by samad-yar-khan** [comment]
@Kamlesh72 how would we handle multiple setting updations leading to multiple sync calls. Do you think this might overload the sync server?

**Comment by Kamlesh72** [comment]
@samad-yar-khan On every /sync request we are aquiring a lock on redis, so multiple sync calls need to wait till that. So I don't think this will overload the sync server.

If you are asking how we will manage multiple settings, then API call will only be made for settings which are updated by user. Once this PR gets merged, my Next PR will be to handle setting changes collectively. https://github.com/middlewarehq/middleware/pull/516#pullrequestreview-2232670225

**Comment by jayantbh** [comment]
@Kamlesh72 were the visual changes that I requested, made? Asking because a bunch of other discussion has happened, and I wonder if I missed it. :) No pressure, just ensuring that I didn't miss anything.

**Comment by Kamlesh72** [comment]
@jayantbh 

https://github.com/user-attachments/assets/2a71e659-ec1d-4e1f-ba38-f9b85cab152a

**Comment by jayantbh** [comment]
Just one tiny change.
Make the confirmation action (Save) be on the left, and the rejection action (Discard) be on the right.

That's generally how accept/reject CTAs work everywhere. :)

Good going on this though! 👍🏽

**Review by jayantbh — APPROVED** [review]

**Comment by jayantbh** [comment]
I've approved it, but I'd like @samad-yar-khan or @adnanhashmi09 to also cross check.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (samad-yar-khan) [inline web-server/src/components/Settings/SyncDaysSetting.tsx:95]: @Kamlesh72 I do not understand the decision, to not allow the user to change default sync days less than current setting. If someone saves 366, I think they should be allowed to update the default sync days to a lower value if they wanted. The minimum value for sync days can be 31 days. Also pls add 366 and 31 days in const variables.
image

**Review by samad-yar-khan — CHANGES_REQUESTED** [review]

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (samad-yar-khan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:95]: also, I see we are not calling sync here, so are we gonna leave it to the cron to sync data on the next sync cycle ?

Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:95]: We can final it after refactor of settings page.

Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:95]: @samad-yar-khan is this fine? I will make the commit.

Comment (samad-yar-khan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:95]: @Kamlesh72 Yup please do. Also, lets call sync after setting updation.

Comment (samad-yar-khan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:95]: It would be super convenient to have data synced as soon as settings are saved, but we will have to see how to handle this incase settings are changed again and again.

Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:95]: @samad-yar-khan as of now, do we need to show some modal to sync?

Comment (samad-yar-khan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:95]: @Kamlesh72 You can ignore my comments about the sync date range, had a talk @jayantbh . He agrees with your logic. The only change I would recommend in this PR is triggering sync after setting change.

Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:95]: @samad-yar-khan We are triggering sync without asking user, right?

Comment (samad-yar-khan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:95]: yes, for now

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (e-for-eshaan) [inline web-server/src/components/Settings/SyncDaysSetting.tsx:139]: Title case

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (e-for-eshaan) [inline web-server/src/components/Settings/SyncDaysSetting.tsx:130]: why double-equals? is this not typesafe?

**Review by e-for-eshaan — CHANGES_REQUESTED** [review]

File: web-server/src/slices/org.ts
Comment (Kamlesh72) [reply web-server/src/slices/org.ts:233]: This is async method, so I used await.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:97]: sentence or capitalize ?

Comment (e-for-eshaan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:97]: Sentence (capitalize first word)

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:76]: I think it is better if label is Capitalized.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (e-for-eshaan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:89]: Updated this comment @Kamlesh72

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:68]: fontSize='20px' no effect, but 'huge' instead of 'big' seems ok.

Comment (e-for-eshaan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:68]: I just wanted to say that instead of using Flexbox, you can easily use <Line>. The props are similar, just use big, small, huge or whichever font size fits the best!

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:52]: If API request fails, we will set Input value to previous Sync days and Show error.

Since updateSyncDays(syncDays) will not throw any error as Redux handles it internally, we can either use then and catch to explicitly throw error [Reference](https://github.com/middlewarehq/middleware/pull/516/files/345fda0a8f57a83f24c81a5b4292b93cc89cfa74#r1723346414).

Or

handle using requestStatus whether request is fulfilled or rejected. I think then even try catch will not be needed.
```js
const handleSave = async () => {
    try {
      const res = await dispatch(
        updateDefaultSyncDaysSettings({ orgId, defaultSyncDays: syncDays })
      ); // This never throws error
      
      if (res.meta.requestStatus === 'rejected') {
        throw new Error();
      }
      enqueueSnackbar('Default Sync Days Updated Successfully', {
        variant: 'success',
        autoHideDuration: 3000
      });
    } catch (e) {
      handleDiscard();
      return enqueueSnackbar('Default Sync Days Update Failed', {
        variant: 'error',
        autoHideDuration: 2000
      });
    }
  };
```

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:130]: Yes it is typesafe. Apologies.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:29]: Got it  🙌🏻 .

Now I see why we shouldn't use payload directly.

syncDays for input and initialSyncDays = useSelector(...)
is this naming ok?

Comment (e-for-eshaan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:29]: lgtm. i would have used `defaultSyncDays`. But not a hard opinion

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (e-for-eshaan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:76]: No hard opinions on that,
Titles can be in title-cases
CTAs, sentences, expressions, facts should be in sentence case.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (e-for-eshaan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:52]: oh, cool!
Just wrap it in a `useCallback`

``` typescript
const func = useCallback(async () => {}, [])
```

Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:52]: @e-for-eshaan If sync days setting is updated successfully we need to call /sync on sync server.

This can be done by calling Nextjs internal API which does force sync, making a entry in UIPreferences table.
```js
handleApi(`/internal/${orgId}/sync_repos`, { method: 'POST' });
```

But I feel instead of force sync, it should be just a normal sync call. But we can't directly call flask sync server API.

Or treating this as force sync is fine?

File: web-server/src/slices/org.ts
Comment (jayantbh) [reply web-server/src/slices/org.ts:233]: I think the `.default_sync_days` property read is only going to happen if the promise resolves successfully, in which case there's functionally no difference between .then or just (await handleApi(...)).default_sync_days.

So this may be a change that can go unchanged.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (jayantbh) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:52]: @Kamlesh72 (unrelated to your last comment)
After the sync setting is updated successfully, I'd recommend making the sync_repos API call on the node server (a.k.a. BFF, a.k.a. nextjs API) instead from the browser.

That way you save time on another request round-trip from the browser to the server.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (jayantbh) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:79]: > You can select how many days of past data Middleware loads. A larger range may increase the initial sync time.

OR

> Select the number of past days to load; more days may increase sync time.

Opinions welcome.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (samad-yar-khan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:52]: @Kamlesh72 why cant we call `syncReposForOrg` after the defauilt sync days setting updation returns a 200 ?

Comment (samad-yar-khan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:52]: This func calls /sync

Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:52]: Screenshot 2024-08-21 at 6 37 43 PM


CHATGPT: The error you're encountering is likely because the Next.js /api routes are server-side only, and the fs module is also server-side (it’s a Node.js module). When you try to call syncReposForOrg from within a React component (which runs on the client side), it attempts to use the fs module, leading to the error.

To resolve this issue, you should ensure that the syncReposForOrg function is only used in server-side contexts or modify it to make a client-side API call instead of directly using the fs module.

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx:79]: Max 75 chars preferable.

How about this?
Select how many days of past data to load; more days may increase sync time.

Comment (e-for-eshaan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx]: kinda detours the convo:
let's go with this:
`Select the number of past days to load; more days may increase sync time.`

Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx]: @e-for-eshaan I have made a commit recently. Any other changes apart from this?

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (e-for-eshaan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx]: `syncReposForOrg` is a server side only function, and should be kept there.
you can call it in any of your API pages

File: web-server/src/components/Settings/SyncDaysSetting.tsx
Comment (e-for-eshaan) [reply web-server/src/components/Settings/SyncDaysSetting.tsx]: once every comment here is resolved, this PR should be good to go. Please comment on this PR when all conversation are resolved and you commit is updated

Comment (Kamlesh72) [reply web-server/src/components/Settings/SyncDaysSetting.tsx]: @e-for-eshaan All comments resolved. PR is updated with given changes.

**Comment by jayantbh** [comment]
image

You guys sure this PR can't use ~12 more comments? 😛

**Comment by jayantbh** [comment]
The initial PR and the PR in its current state is significantly different.
Great growth @Kamlesh72 👏🏽 
This must have been something to persevere through, but great job sticking to it.

**Comment by Kamlesh72** [comment]
Thanks a lot everyone! Your feedback and guidance have made all the difference.
The learnings I got is invaluable.

**Review by samad-yar-khan — APPROVED** [review]

**Comment by samad-yar-khan** [comment]
LGTM @Kamlesh72 Great Work 🚀

**Comment by e-for-eshaan** [comment]
LGTM @Kamlesh72 , good stuff! Merging this 🎉 
cc: @samad-yar-khan @adnanhashmi09 @jayantbh

**Review by e-for-eshaan — APPROVED** [review]

**Comment by jayantbh** [comment]
Aw shucks. 93 (94 with this) comments. We were 6 away from greatness. 😂

**Pull request merged**
